### DISCUSSION TITLE
Linear Programming with GLPK

### DISCUSSION THREAD

**Discussion by michaels0m** [cvxpy/cvxpy#1380] [Q&A] [answered]
Hi, I have been mainly working with Linear Programming problems which can get quite large (10,000+ variables along with 30,000+ linear constraints) and I have found the GLPK solver quite useful for solving these problems, however I have a couple of questions with regards to how it is currently implemented in CVXPY:

1) Going through the source code I notice that CVXPY accesses the GLPK solver via CVXOPT. Are there any transformations occurring before the problem is being passed to CVXOPT in order to improve the format of the problem e.g. scaling or reductions? If not, would it be quicker to use the GLPK solver via CVXOPT?

2) I have noticed that GLPK has several methods for solving LPs including glpk_simplex, glpk_exact and glpk_interior (which can be found in the documentation attached). Am I right in assuming that the default is glpk_simplex? If so, is there a way of accessing the glpk_exact and glpk_interior methods?
[glpk_documentation.pdf](https://github.com/cvxpy/cvxpy/files/6511718/glpk_documentation.pdf)

**Comment by rileyjmurray** [comment, marked as answer]
Hi @Michael-git96.

For your first question: CVXPY makes no real attempt to improve the format of the problem before passing to CVXOPT or GLPK. We check for some degenerate inputs to prevent certain solver errors, but nothing more than that. Definitely no scaling or presolve. There's a good chance you'd see slightly faster performance calling GLPK directly from CVXOPT. It's really a matter of whether CVXOPT or CVXPY compiles symbolic expressions into matrix data faster, and I think that's going to be problem-dependent.

For your second question: CVXOPT's interface to GLPK is defined by [this file](https://github.com/cvxopt/cvxopt/blob/60fdb838e0bb2d8f32ba51129552c83b55acd2a7/src/C/glpk.c). You'll notice that it only supports ``glpk_simplex``. There's probably not much value in exposing ``glpk_interior``, since differences in IPM's for LP are mostly matters of linear algebra efficiency. Page 64 of the GLPK documentation you linked shows that GLPK doesn't have fancy linear algebra routines, while CVXOPT has several pre-defined linear algebra routines that you can select from based on problem structure (granted it doesn't make this decision for you, but the option is there). There is potential value in exposing ``glpk_exact``, but that brings up the issue that CVXPY (and CVXOPT) only support double-precision floating-point numbers, while ``glpk_exact`` performs computations with arbitrary precision rational numbers.

**Comment by michaels0m** [comment]
Hi @rileyjmurray,

Thank you for your reply. I read through the documentation on GLPK and I noticed that the IPM was missing a few things which are supposed to ensure that it terminates once it has found the optimal point. I actually upgraded the scipy version I was using and noticed that they have introduced the HiGHS solvers (https://www.maths.ed.ac.uk/hall/HiGHS/#guide) into the linprog() function (https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html). This was particularly useful for me as I use scipy.sparse matrices quite often due to memory restrictions and the constraint matrices usually being quite sparse. The performance was comparable to the performance of GLPK on sparse LPs. Are there any plans to integrate the HiGHS solvers into CVXPY?

**Reply by rileyjmurray** [reply]
@Michael-git96 I had no idea SciPy had access to such a nice implementation! We would love it if someone would contribute an interface to HiGHS to CVXPY, by wrapping the SciPy interface. I imagine there is little to no overhead by calling from the SciPy interface since it requires input in terms of pure matrices and vectors. Would you be up to trying to contribute such an interface? If so I can provide you with advice and pointers.

**Reply by michaels0m** [reply]
I would be happy to contribute and try to wrap the SciPy interface. I would greatly appreciate any advice and pointers you would have!

**Reply by michaels0m** [reply]
@rileyjmurray - I have managed to clone the repo and follow the steps in the CONTRIBUTING.md file (https://github.com/cvxpy/cvxpy/blob/master/CONTRIBUTING.md#code-style) including running setup.py. I took a quick look at the code but it is quite a large project. Do you have any advice on which areas/scripts to take a look at first? Also if I am not mistaken, it seems that when the cvxpy.Problem() instance is initialised, the constraint vectors and matrices are immediately transformed from their original format (NumPy or Scipy) into a different format?

**Reply by SteveDiamond** [reply]
@Michael-git96 not sure if you saw this: https://www.cvxpy.org/contributing/index.html#solver-interfaces. The best place to start is by looking at existing interfaces, for example ECOS: https://github.com/cvxpy/cvxpy/blob/master/cvxpy/reductions/solvers/conic_solvers/ecos_conif.py

**Reply by michaels0m** [reply]
@SteveDiamond thank you for the link for contributing towards a new solver-interface, I only read the CONTRIBUTING.md file which was less detailed. I will take it step by step and start building an interface to scipy linprog (scipy_conif.py) so that CVXPY will have access to scipy's revised-simplex, interior-point methods as well as the HiGHS solvers. I just had a couple of questions:
1) If I have any questions, should they be communicated in this discussion section?
2) Is there a specific CVXPY principal developer which I should be coordinating with for this new interface?

**Reply by SteveDiamond** [reply]
You can discuss in our discord server (https://discord.gg/zZvBtT7S).

**Reply by rileyjmurray** [reply]
@Michael-git96 for context, we just created the Discord server, so we haven't really settled on norms about how it use it in the development process. If you want, you can create a GitHub issue for adding an interface to SciPy-accessible LP solvers. The advantage there is that you can easily reference our codebase and any discussions that we've had elsewhere on GitHub.

**Reply by michaels0m** [reply]
@rileyjmurray - Thank you for the advice, I have just created a new GitHub issue for the SciPy interface as I have already started to work on it and I had a couple of questions.
